### PR TITLE
Add models operations to twin lifecycle sample

### PR DESCRIPTION
**Pull request by abhipsaMisra**
Add models operations to twin lifecycle sample

### CODE REVIEW

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java [hunk 1]
```diff
@@ -909,7 +909,7 @@ public HttpPipeline getHttpPipeline() {
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Void> deleteModel(String modelId) {
         return deleteModelWithResponse(modelId)
-            .map(Response::getValue);
+            .flatMap(voidResponse -> Mono.empty());
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:912]: `.map(Response::getValue)` returns null, which is not a valid return item from a Mono. So we need to map it to `Mono.empty()` instead.

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 2]
```diff
@@ -88,108 +98,250 @@ public static void main(String[] args) throws IOException, InterruptedException
...
+        // Once the async API terminates (either successfully, or with an error), the latch count is decremented, or the semaphore is released.
+        for (Map.Entry<String, String> twin : twins.entrySet()) {
+            String twinId = twin.getKey();
+
+            List<BasicRelationship> relationshipList = new ArrayList<>();
+            Semaphore listRelationshipSemaphore = new Semaphore(0);
+            Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
+            CountDownLatch deleteTwinsLatch = new CountDownLatch(1);
+
+            // Call APIs to retrieve all relationships.
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:148]: This implementation is very long and verbose. I am trying to see how I could chain the list and delete operations together, I'll put up an update in the next PR.

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 3]
```diff
@@ -88,108 +98,249 @@ public static void main(String[] args) throws IOException, InterruptedException
...
+     * @throws IOException If an I/O error is thrown when accessing the starting file.
+     * @throws InterruptedException If the current thread is interrupted while waiting to acquire latch.
+     */
+    public static void createAllModels() throws IOException, InterruptedException {
+        System.out.println("CREATING MODELS");
+        List<String> modelsToCreate = new ArrayList<>(FileHelper.loadAllFilesInPath(ModelsPath).values());
+        final CountDownLatch createModelsLatch = new CountDownLatch(1);
+
+        // Call API to create the models. For each async operation, once the operation is completed successfully, a latch is counted down.
+        client.createModels(modelsToCreate)
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:247]: Q - is `.createModels()` an create-all-or-one operation?

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 4]
```diff
@@ -88,108 +98,249 @@ public static void main(String[] args) throws IOException, InterruptedException
...
+     * @throws InterruptedException If the current thread is interrupted while waiting to acquire latch.
+     */
+    public static void deleteAllModels() throws InterruptedException {
+        System.out.println("DELETING MODELS");
+
+        // This is to ensure models are deleted in an order such that no other models are referencing it.
+        List<String> models = asList(RoomModelId, WifiModelId, BuildingModelId, FloorModelId, HvacModelId);
+
+        // Call APIs to delete the models.
+        // Note that we are blocking the async API call. This is to ensure models are deleted in an order such that no other models are referencing it.
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:222]: We should have been able to schedule the async API on a single thread, and forced them to run sequentially; however, I didn't have success with that.
I am blocking the async API call for now, until I figure that out.

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 5]
```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
...
+                // Now the digital twin should be safe to delete
+
+                // Call APIs to delete the twins.
+                client.deleteDigitalTwin(twinId)
+                    .doOnSuccess(aVoid -> System.out.println("Deleted digital twin: " + twinId))
+                    .doOnError(IgnoreNotFoundError)
+                    .doOnTerminate(deleteTwinsLatch::countDown)
+                    .subscribe();
+
+                // Wait until the latch has been counted down for each async delete operation, signifying that the async call has completed successfully.
```
Comment (bikamani) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:191]: For each async delete? Can you elaborate what you mean here?

Comment (drwill-ms) [reply sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:191]: What is a latch?

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 6]
```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
...
+                .doOnTerminate(listRelationshipSemaphore::release)
+                .subscribe();
+
+            // Call APIs to delete all relationships.
+            if (listRelationshipSemaphore.tryAcquire(2, MaxWaitTimeAsyncOperationsInSeconds, TimeUnit.SECONDS)) {
+                relationshipList
+                    .forEach(relationship -> client.deleteRelationship(relationship.getSourceId(), relationship.getId())
+                        .doOnSuccess(aVoid -> {
+                            if (twinId.equals(relationship.getSourceId())) {
+                                System.out.println("Found and deleted relationship: " + relationship.getId());
```
Comment (drwill-ms) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:170]: Does Java support string interpolation?

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java (hunk 5 above)
Comment (drwill-ms) [reply sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:191]: Appears to be some kind of threadsafe counter, eh?

I'd reword this to be "Wait until the latch count reaches zero..."

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 7]
```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
...
-                                throwable -> System.err.println("Delete incoming relationship error: " + throwable)
-                            ));
+        // Once the async API terminates (either successfully, or with an error), the latch count is decremented, or the semaphore is released.
+        for (Map.Entry<String, String> twin : twins.entrySet()) {
+            String twinId = twin.getKey();
+
+            List<BasicRelationship> relationshipList = new ArrayList<>();
+            Semaphore listRelationshipSemaphore = new Semaphore(0);
+            Semaphore deleteRelationshipsSemaphore = new Semaphore(0);
+            CountDownLatch deleteTwinsLatch = new CountDownLatch(1);
```
Comment (drwill-ms) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:148]: Unfortunate that you have to know ahead of time how many items you are counting down. This presents a "magic" number, which can be confusing to someone else reviewing the code who doesn't realize the significance (this will be used to asynchronously delete 1 dt).

Also, in C# we have a general rule of not declaring a variable until it is needed (so it is more contextual). Does Java have a different standard of all variables at the top?

FWIW, I think this could at least use a comment.

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java (hunk 5 above)
Comment (abhipsaMisra) [reply sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:191]: A countdown latch is a mechanism to block the calling thread until other threads that are running in parallel have completed (counted down). The `.countdown()` essentially decrements a thread-safe counter.

From this article online: https://www.baeldung.com/java-countdown-latch
Simply put, a CountDownLatch has a counter field, which you can decrement as we require. We can then use it to block a calling thread until it's been counted down to zero.

If we were doing some parallel processing, we could instantiate the CountDownLatch with the same value for the counter as a number of threads we want to work across. Then, we could just call countdown() after each thread finishes, guaranteeing that a dependent thread calling await() will block until the worker threads are finished.

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java [hunk 8]
```diff
@@ -88,108 +100,220 @@ public static void main(String[] args) throws IOException, InterruptedException
...
-                                } else {
-                                    System.err.println("Could not delete digital twin " + twinId + " due to " + throwable);
-                                }
-                            })
-                            .subscribe();
+                    client.deleteModel(modelId).block();
+                    System.out.println("Deleted model: " + modelId);
+                } catch (ErrorResponseException ex) {
+                    if (ex.getResponse().getStatusCode() != HttpStatus.SC_NOT_FOUND) {
+                        System.err.println("Could not delete model " + modelId + " due to " + ex);
```
Comment (drwill-ms) [inline sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:216]: If Java doesn't have string interpolation, I'd think this would be much more readable as a string format (like below).

**Review by drwill-ms — APPROVED** [review]

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java (hunk 6 above)
Comment (abhipsaMisra) [reply sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:170]: We can use `String.format()` to specify a template, but nothing as handy as `$` in C#.

**Review by azabbasi — COMMENTED** [review]
I think this sample is doing way too much work. Reconsider making it just creating a brand new model with brand new id and create a twin off of that and then clean up after itself. 
That would make it concise and easier to navigate.

**Review by azabbasi — APPROVED** [review]

File: sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java (hunk 7 above)
Comment (abhipsaMisra) [reply sdk/digitaltwins/azure-digitaltwins-core/src/samples/java/com/azure/digitaltwins/core/DigitalTwinsLifecycleAsyncSample.java:148]: We need to initialize these latches and semaphore within the scope that they are referenced in; but yes, I agree, the count can be confusing to understand.
I'll add some more comments around this.
